src/machine.rs

```diff
@@ -237,32 +247,35 @@ impl<'a, 'tcx> Machine<'a, 'tcx> {
         self.stack.last_mut().unwrap()
     }
 
+    pub fn get_or_insert_func(&mut self, def_id: DefId) -> &mut Rc<Function<'tcx>> {
+        self.funcs.entry(def_id).or_insert({
+            let instance = Instance::mono(self.tcx, def_id).def;
+            let mir = match instance {
+                InstanceDef::Item(def_id) => {
+                    self.tcx.maybe_optimized_mir(def_id).unwrap()
+                },
+                _ => self.tcx.instance_mir(instance),
+            };
+            Rc::new(Function::new(def_id, mir))
+        })
+    }
+
     pub fn push_frame(&mut self,
                       def_id: DefId,
                       args: Vec<TyVal<'tcx>>,
                       ret_val: Option<Address>,
-                      ret_block: Option<BasicBlock>){
-        let def = Instance::mono(self.tcx, def_id).def;
-        let mir = self.load_mir(def);
-
-        // Calculate local indexes
-        let mut local_idxs: Vec<usize> = Vec::with_capacity(mir.local_decls.len());
-        let mut next_idx = 0;
-        for decl in mir.local_decls.iter() {
-            local_idxs.push(next_idx);
-            next_idx += size_of(decl.ty);
-        }
-
+                      ret_block: Option<BasicBlock>) {
+        let func = self.get_or_insert_func(def_id).clone();
         // Init locals with args and zero the rest
-        let mut locals: Vec<u8> = vec![0; next_idx];
+        let mut locals: Vec<u8> = vec![0; func.locals_size];
         let args: Vec<u8> = args.into_iter().flat_map(|x| x.to_bytes()).collect();
         locals.splice(0..args.len(), args.into_iter());
 
-        let frame = Frame::new(def_id, mir, local_idxs, locals, ret_val, ret_block);
-        self.stack.push(frame)
+        let frame = Frame::new(func.clone(), locals, ret_val, ret_block);
```

Comment:
Conventionally I think this should be `Rc::clone(func)`.

Reply:
Makes sense. To make it more explicit that you're cloning the reference and not the data?

Reply:
Yes.